nixos/lib/test-driver/test-driver.py

```diff
@@ -1002,19 +1002,24 @@ def subtest(name: str) -> Iterator[None]:
     ]
     exec("\n".join(machine_eval))
 
-    @atexit.register
     def clean_up() -> None:
         with log.nested("cleaning up"):
             for machine in machines:
                 if machine.pid is None:
                     continue
                 log.log("killing {} (pid {})".format(machine.name, machine.pid))
+                machine.send_monitor_command("quit")
```

Comment:
Just to confirm: if you omit the change on this line it does not fix the problem?

Reply:
That line shouldn't actually make any difference.
The line just below is issuing a `SIGKILL` (`-9`) to qemu.

Reply:
What makes a difference is the `serial_thread.join()` which ensure the thread responsible for reading the qemu output has exited before we exit the script. That was the racy behavior we were observing.

Reply:
What is the reason for the `quit` command here?

Reply:
(we need to ensure qemu is stopped and that the pipe is going to EOF before calling the `serial_thread.join()`)

Reply:
> What is the reason for the `quit` command here?

mostly because I was trying to follow the logic of the upstream code.